Altimeter/Src/altimeter.cpp

```diff
@@ -0,0 +1,229 @@
+#include <altimeter.hpp>
+
+MS5611::MS5611(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, GPIO_TypeDef *ps_port, uint16_t cs_pin, uint16_t ps_pin){
+	spi_handle_ = spi_handle;
+	cs_port_ = cs_port;
+	ps_port_ = ps_port;
+	cs_pin_ = cs_pin;
+	ps_pin_ = ps_pin;
+	HAL_GPIO_WritePin(ps_port_, ps_pin_, GPIO_PIN_RESET);
+
+	setup_done_ = true;
+	base_elevation_set_ = true;
+	getProm();
+
+	// Grab the sensor's altitude above sea level while on some surface and assign it to base_elev_
+	// Subtract this from measurements at greater heights for the height above that initial surface.
+	for (int i = 0; i < 100; i++) { //take a few measurements to make sure they are accurate/consistent.
+		setBaseElevation();
+	}
+}
+
+void MS5611::reset(){
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_RESET);
+
+	uint8_t tx_data[3] = {RESET_COMMAND, 0xFF, 0xFF};
+	uint8_t rx_data[3];
+	uint16_t data_size = 3;
+
+	HAL_StatusTypeDef spi_status_ = HAL_SPI_TransmitReceive_IT(spi_handle_, tx_data, rx_data, data_size);
+	if (spi_status_ != HAL_OK) {
+		setup_done_ = false;
+	}
+
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_SET);
+	osDelay(3);
+}
+
+uint16_t MS5611::getPromElement(uint8_t address){
+
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_RESET);
+
+	uint8_t tx_data[3] = {address, 0xFF, 0xFF};
+	uint8_t rx_data[3];
+	uint16_t data_size = 3;
+
+	HAL_StatusTypeDef spi_status_ = HAL_SPI_TransmitReceive_IT(spi_handle_, tx_data, rx_data, data_size);
+	if (spi_status_ != HAL_OK) {
+		setup_done_ = false;
+	}
+
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_SET);
+
+	// construct 16-bit calibration coefficient.
+	uint16_t prom_element = ((rx_data[1] << 8) | (rx_data[2])) & 0x0000FFFF;
+
+	return prom_element;
+}
+
+void MS5611::getProm(){
+	reset();
+	for (int i = 0; i < 8; ++i) {
+		ms6511_prom_data_[i] = getPromElement(PROM_READ_ADDRESS_0 + 2 * i);
+		if (!setup_done_){
+			return;
+		}
+	}
+
+	setup_done_ = true;
+}
+
+bool MS5611::getBarometer(uint8_t conversion_command, uint32_t &uncompensated_pressure_temperature){
+
+	if(!setup_done_){
+		return false;
+	}
+
+	uint8_t tx_data = conversion_command;
+	uint8_t rx_data;
+	uint8_t tx_data_2[4] = {ADC_READ, 0xFF, 0xFF, 0xFF};
+	uint8_t rx_data_2[4];
+	uint16_t data_size = 1;
+
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port_, cs_pin_, GPIO_PIN_RESET);
+
+	// send the digital pressure or temperature with the command.
+	HAL_StatusTypeDef spi_status_ = HAL_SPI_TransmitReceive_IT(spi_handle_, &tx_data, &rx_data, data_size);
```

Comment:
Sorry, it was me in the previous PR saying maybe we could use interrupt-based data transceiving in the future, but I actually meant using DMA. And here I think it makes sense to use normal polling since you are reading the transmitting at the same time. I thought if you call the hal function ended with `_IT`, you are also triggering an ISR callback function, but you also don't need to use it here, so there is not a big difference between using interrupt base or pulling based transceiving.